Font family: switch from CSS Custom Property to classes

ockham opened the pull request:
Fixes #31489

This PR switches the font-family property to use class names instead of CSS Custom Properties, following what the other properties do.

## How has this been tested?

- Use an FSE-enabled theme (e.g. TT1 Blocks).
- Create a new post, and insert a block that has support for the font-family style enabled (e.g. Verse).
- In the block sidebar's typography panel, change the font family.
- Verify that the block's font changes accordingly.
- Check the Code Editor. Verify that it matches the below snippet.
- Exit the Code Editor.
- Publish the post, and view it in the front end. Verify that the font family class name is applied to the frontend.

## Block markup changes

### Before

```html
<!-- wp:verse {"style":{"typography":{"fontFamily":"var:preset|font-family|geneva-verdana"}}} -->
<pre class="wp-block-verse" style="font-family:var(--wp--preset--font-family--geneva-verdana)">Font family</pre>
<!-- /wp:verse -->
```

### After

```html
<!-- wp:verse {"fontFamily":"geneva-verdana"} -->
<pre class="wp-block-verse has-geneva-verdana-font-family">Font family</pre>
<!-- /wp:verse -->
```

## Screencasts

### Before

![font-family-before](https://user-images.githubusercontent.com/96308/118710163-37d27080-b81e-11eb-889b-aac3a1ba3791.gif)

### After

![font-family-after](https://user-images.githubusercontent.com/96308/118710242-52a4e500-b81e-11eb-964d-4bf09112db0d.gif)

## Tasks

- [x] Update server hook.
- [x] Update client hook.
- [x] Tell the engine to generate the font-family CSS Custom Properties and classes.
- [x] Update tests.
- [x] Add deprecations to move the font-family information from `style.typography.fontFamily` into a top-level `fontFamily` attribute.
  - [x] Button
  - [x] List
  - [x] Navigation
  - [x] Post Comment Author
  - [x] Post Comment Date
  - [x] Post Date
  - [x] Post title
  - [x] Query title
  - [x] Site tagline
  - [x] Site title
  - [x] Verse

ockham committed:
Client hooks: use class names for font family

ockham commented:
Quick check-in with @nosolosw (before I start working on migrations): Are we okay with introducing a new top-level `fontFamily` attribute here (rather than keeping the nested `style.typography.fontFamily`)? (See block markup examples in the PR desc.)

It seemed to me that `style.typography.fontFamily` wouldn't make much sense with font family slugs (such as `geneva-verdana`). If we would like to keep the information in that nested attribute, we should probably spell out the font family (e.g. `Geneva, Tahoma, Verdana, sans-serif`)?

Furthermore, is it our responsibility to translate the class to a style on the frontend, or is it the theme's to style it accordingly? (I guess it's ours?)

Curious to hear your thoughts.

ockham commented on lib/block-supports/typography.php:
```diff
@@ -89,18 +89,19 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
 				$font_family_name = substr( $font_family, $index_to_splice );
 				$styles[]         = sprintf( 'font-family: var(--wp--preset--font-family--%s);', $font_family_name );
 			} else {
-				$styles[] = sprintf( 'font-family: %s;', $block_attributes['style']['color']['fontFamily'] );
```
This looks like a bug to me -- I assume `color` should be `typography`?

oandregal replied:
I asked at https://github.com/WordPress/gutenberg/pull/24868/files#r662946264 but I presume this is a code path that is never reached.

oandregal replied:
Actually, it looks like there's a legit use case: patterns that use a font directly instead of the slug.

ockham commented on lib/block-supports/typography.php:
```diff
@@ -89,18 +89,19 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
+				$styles[] = sprintf( 'font-family: %s;', $font_family );
 			}
+		} elseif ( isset( $block_attributes['fontFamily'] ) ) { // Current attribute style.
+			$styles[] = sprintf( 'font-family: %s;', $block_attributes['fontFamily'] );
```
I still must be doing something wrong, I don't see this applied to the Verse block on the frontend 🤔

ockham committed:
Server supports: use class

oandregal committed:
theme.json: enqueue classes for font-family

oandregal committed:
Remove unnecessary code path

oandregal commented on lib/block-supports/typography.php:
```diff
@@ -105,16 +105,17 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
+
+		if ( $has_named_font_family ) {
+			$classes[] = sprintf( 'has-%s-font-family', $block_attributes['fontFamily'] );
+		} elseif ( $has_legacy_font_family ) {
```
We should be able to remove this path in a couple of plugin versions.

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -191,7 +191,12 @@ class WP_Theme_JSON_Gutenberg {
-			'classes'       => array(),
+			'classes'       => array(
+				array(
+					'class_suffix'  => 'font-family',
```
The engine generates the font-family classes.

oandregal committed:
Do not use inline styles

We don't want to serialize font-family to inline styles.
We did so for other presets (font-size, color) because
themes forgot to add the corresponding classes.

oandregal committed:
Add class name in edit

oandregal commented:
Worked a bit on this today and this is working. Tasks left are: block deprecations + update tests.

oandregal committed:
Support custom font families

There may be cases in which patterns add their own font-families
without being slugs for us to process.

oandregal committed:
Update tests

oandregal commented:
I've picked up this today and updated the tests. The remaining step is to add block deprecations.

oandregal committed:
Add deprecation for the button block

oandregal committed:
clean empty object extracted to utils

oandregal committed:
Add deprecation for the list block

oandregal committed:
Add deprecation for the post title block

oandregal committed:
Add deprecation for the post title block

oandregal committed:
fixup site-title

oandregal committed:
Add deprecation for the verse block

oandregal committed:
Add deprecation for the site-tagline block

oandregal committed:
Add deprecation for the navigation block

oandregal committed:
Add deprecation for the query-title block

oandregal commented:
I've pushed the last bits necessary to land this, and it's ready.

oandregal commented on lib/block-supports/typography.php:
```diff
@@ -105,16 +105,21 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
+			$classes[] = sprintf( 'has-%s-font-family', $block_attributes['fontFamily'] );
+		} elseif ( $has_custom_font_family ) {
+			// Before using classes, the value was serialized as a CSS Custom Property.
+			// We don't need this code path it lands in core.
```
```suggestion
			// We don't need this code path when it lands in core.
```

oandregal committed:
Update lib/block-supports/typography.php

oandregal commented:
Rebased this from `trunk` due to some conflicts. Any @WordPress/gutenberg-core can help land this? Perhaps @jorgefilipecosta or @aristath as you were involved in the original PR?

oandregal committed:
Update test after rebase

oandregal committed:
Update test after rebase

oandregal commented:
Rebased to fix conflicts from `trunk`.

ockham commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -73,6 +80,31 @@ const migrateCustomColorsAndGradients = ( attributes ) => {
+		fontFamily,
+	};
+};
+
```
Since this is now copied across eight different blocks, I think it'd make sense to extract it into a little helper file in `packages/block-library/src/utils/`. That file could be called `migrations.js` and (later) also hold other shared migrations..

oandregal replied:
Pushed that change at https://github.com/WordPress/gutenberg/pull/31910/commits/062e4e243f559192924c32b85733b518a159828e

I thought about that but hadn't seen any prior art of migrations being shared across blocks, despite some of them be actually shared. So I presumed the consensus was to have the migrations file be isolated. Happy to change gears.

ockham replied:
> Pushed that change at [062e4e2](https://github.com/WordPress/gutenberg/commit/062e4e243f559192924c32b85733b518a159828e)

Thanks!

> I thought about that but hadn't seen any prior art of migrations being shared across blocks, despite some of them be actually shared. So I presumed the consensus was to have the migrations file be isolated. Happy to change gears.

Yeah, I noticed that too, but I thought it was out of habit rather than consensus -- it just grew that way. Plus there might still not be that many examples of shared migrations, since those arise mostly out of the block-supports mechanism, when updated. Anyway, that's at least how I rationalized it 😅  Thanks for changing, I do believe it's the right thing to do 😄

ockham replied:
Oh, one more ask: Let's rename `oldFontFamilyMigration` to `migrateFontFamily`. The reason for this is that I think we have a precedent in migrations set by `oldColorsMigration`, which is indeed an even _older_ migration than the "current" one (which is called `migrateCustomColorsAndGradients`). To elaborate:

`migrateCustomColorsAndGradients` transforms `customTextColor`, `customBackgroundColor`, and `customGradient` into their respective `style.color.*` counterparts (which is the current way of storing these attributes).

`oldColorsMigration` OTOH first transforms the even older `textColor` and `color` attributes into `customTextColor` and `customColor`, respectively, and _then_ applies `migrateCustomColorsAndGradients` to transform those into their `style.color.*` counterparts.

ockham replied:
Renamed in 541573c6ca3226812fa698515b270bc3fca80a72.

oandregal committed:
Extract font family migration to file

ockham commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -841,6 +873,144 @@ const deprecated = [
+				</div>
+			);
+		},
+		migrate: oldFontFamilyMigration,
```
I'm never quite sure if we only need the 'latest' migration here, or if we need to compose _all_ migrations, since only _one_ deprecation will be run.

In the past, I think I've done the latter, see e.g. https://github.com/WordPress/gutenberg/pull/30194/files#diff-e133a7448e28a248ec35cb9a2943bb23e00755612533ac29e015532a6d618f9eR416-R419

One way to test would be to compose button markup for a button that has the 'old' font family attribute, and e.g. the 'old' `borderRadius`, paste that into the Code Editor, and switch to the Visual Editor and back to the Code Editor to see if it migrates properly. Something like

```html
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"borderRadius":25,"textColor":"white","style":{"typography":{"fontFamily":"var:preset|font-family|hoefler-times-new-roman"}}} -->
<div class="wp-block-button" style="border-radius:25px; font-family:var(--wp--preset--font-family--hoefler-times-new-roman)"><a class="wp-block-button__link has-white-color has-text-color">Button</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
``` 

But alas, this even fails to properly migrate on `trunk` 🤔

ockham replied:
Let's maybe at least add some font family deprecation test coverage to `test/integration/fixtures/blocks`, e.g. something like `core__button__font_family__deprecated.html` and friends 😄

ockham replied:
> Let's maybe at least add some font family deprecation test coverage to `test/integration/fixtures/blocks`, e.g. something like `core__button__font_family__deprecated.html` and friends 😄

f58cfa9

ockham replied:
Unthreading the convo on composed migrations here: https://github.com/WordPress/gutenberg/pull/31910#issuecomment-943560377

ockham committed:
Add some deprecation coverage for the Button block

ockham commented:
Following up from https://github.com/WordPress/gutenberg/pull/31910#discussion_r728881356:

Yeah, so it seems like the migration is indeed incomplete. E.g. when providing this fixture (based on https://github.com/WordPress/gutenberg/commit/f58cfa9db2766f9fc0b65870d256bd25751864ee, with an added `borderRadius` attribute)

```html
<!-- wp:button {"borderRadius":25,"style":{"typography":{"fontFamily":"var:preset|font-family|cambria-georgia"}}} -->
<div class="wp-block-button" style="font-family:var(--wp--preset--font-family--cambria-georgia)"><a class="wp-block-button__link" style="border-radius:25px">My button</a></div>
<!-- /wp:button -->
```

I get this result (in the `*.serialized.html` file):

```html
<!-- wp:button {"style":{"typography":{"fontFamily":"var:preset|font-family|cambria-georgia"},"border":{"radius":"25px"}}} -->
<div class="wp-block-button" style="font-family:var(--wp--preset--font-family--cambria-georgia)"><a class="wp-block-button__link" style="border-radius:25px">My button</a></div>
<!-- /wp:button -->
```

Note that the border radius migration is performed, but the font family one isn't 😕 

I think that this is in the nature of deprecations/migrations: Only one of them is performed. So IIUC, we'd need to update _all_ deprecations in that file to now include the font family migration.

This seems rather unwieldy, and like a bigger problem with the way migrations work; plus I bet that the same issue exists for many of our existing migrations.

I think I'll file an issue to discuss this at large (as it makes writing and reviewing deprecations incredibly error-prone and hard).

I'm not sure what to do in the short term for this PR. I'd rather not close my eyes on the issue, since it _will_ break users' content (it's only debatable how badly -- most likely, their buttons will use the font family setting if they have other customizations applied). Maybe we can after all add the new migration to all existing ones 🤔

ockham committed:
Rename oldFontFamilyMigration to migrateFontFamily

ockham commented:
> I think I'll file an issue to discuss this at large (as it makes writing and reviewing deprecations incredibly error-prone and hard).

https://github.com/WordPress/gutenberg/issues/35647

oandregal committed:
List: add fixtures and update deprecation list

oandregal committed:
Post Title: update deprecation list

oandregal committed:
Post title: add fixtures

ockham commented on packages/block-library/src/list/deprecated.js:
```diff
@@ -0,0 +1,77 @@
+		},
+		__unstablePasteTextInline: true,
+		__experimentalSelector: 'ol,ul',
+	},
```
I'm now thinking a bit more about structuring deprecations better and less redundant 🤔 

How about we do

```suggestion
import blockConfig from './block.json';

const {
	attributes: currentAttributes,
	supports: currentSupports,
} = blockConfig;

const v1 = {
	attributes: currentAttributes,
	supports: currentSupports,
```

This pattern should work for all blocks deprecations that update to the block's current attributes format (and block supports), I think.

ockham replied:
Trying out in b60ba0e998. We can revert if needed.

ockham commented on packages/block-library/src/list/deprecated.js:
```diff
@@ -0,0 +1,77 @@
+		__unstablePasteTextInline: true,
+		__experimentalSelector: 'ol,ul',
+	},
+	save( { attributes } ) {
+		const { ordered, values, type, reversed, start } = attributes;
+		const TagName = ordered ? 'ol' : 'ul';
+
+		return (
+			<TagName { ...useBlockProps.save( { type, reversed, start } ) }>
+				<RichText.Content value={ values } multiline="li" />
+			</TagName>
+		);
+	},
```
Similar here, we could

```js
import currentSave from './save.js';
```

and do

```suggestion
	save: currentSave,
```

here...

ockham replied:
Trying out in b60ba0e998. We can revert if needed.

ockham commented on packages/block-library/src/utils/migrate-font-family.js:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Internal dependencies
+ */
+import cleanEmptyObject from '../utils/clean-empty-object';
```
```suggestion
import cleanEmptyObject from './clean-empty-object';
```

ockham committed:
Express list deprecation through current block.json and save

ockham committed:
Express PostTitle block deprecation attributes through block.json

ockham committed:
Simplify import

ockham committed:
Express Site Tagline block deprecation attributes through block.json

ockham committed:
Version Site Tagline deprecations

ockham committed:
Add Site Tagline fixtures

ockham committed:
Express Site Title block deprecation attributes through block.json

ockham committed:
Version Site Title deprecations

ockham committed:
Add Site Title fixtures

ockham committed:
Express Query Title block deprecation attributes through block.json

ockham committed:
Version Query Title deprecations

ockham committed:
Add Query Title fixtures

ockham committed:
Version Verse block deprecations

ockham committed:
Verse deprecation: Inline v1 block attributes

ockham committed:
Express Verse block v2 deprecation through block.json and save

ockham committed:
Rename Verse deprecation fixture

ockham committed:
Add fixtures for Verse block deprecation v2

ockham committed:
Express Button block latest deprecation through block.json and save

ockham committed:
Express Navigation block latest deprecation through block.json and save

ockham committed:
Add fixtures for Navigation block font family deprecation

ockham commented:
I’ve added fixtures for all blocks affected by the font family attribute migration, and expressed the deprecations for all blocks through their current attributes, supports, and (if applicable), `save` function (see https://github.com/WordPress/gutenberg/pull/31910#discussion_r729795487 and https://github.com/WordPress/gutenberg/pull/31910#discussion_r729797607 for more information on this). I think that this should make deprecations both stabler and easier to review. I plan to update `block-deprecations.md` in order to recommend this as a best practice (in a separate PR).

For most blocks touched here, this is fortunately the first deprecation. Only the Button, Navigation, and Verse blocks have seen deprecations before, and Verse's didn’t even have a migration.

This means that we now have to go through Button's and Navigation's deprecations and potentially compose their existing migrations with `migrateFontFamily`.

ockham commented:
On another note, it seems like the latest deprecation should be added at the _top_ of the deprecations array (rather than the bottom). This is e.g. reflected in the Navigation block deprecations (where the latest deprecation prior to this one was just added to the top), and [documented in `block-deprecations.md`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/#example).

> The deprecations in the array should be in reverse chronological order. This allows the block editor to attempt to apply the most recent and likely deprecations first, avoiding unnecessary and expensive processing.

ockham committed:
Move latest Navigation block deprecation to the top

ockham committed:
Move latest Verse block deprecation to the top

ockham committed:
Move latest Button block deprecation to the top

ockham commented:
> On another note, it seems like the latest deprecation should be added at the _top_ of the deprecations array (rather than the bottom). This is e.g. reflected in the Navigation block deprecations (where the latest deprecation prior to this one was just added to the top), and [documented in `block-deprecations.md`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/#example).
> 
> > The deprecations in the array should be in reverse chronological order. This allows the block editor to attempt to apply the most recent and likely deprecations first, avoiding unnecessary and expensive processing.

Done per the latest three commits 🙂

oandregal commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -16,6 +16,18 @@ import {
+import blockConfig from './block.json';
+import currentSave from './save';
+
+const {
```
Hey, I read through the [deprecations code](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/parser/apply-block-deprecated-versions.js#L56), and I don't think we should do this.

These items aim to represent the deprecation items. It's an accident that it works in this particular case. It works because attribute definition is coincidentally the same: the `style` attribute of type object is still defined (implicitly, by the hook) in both cases, although its internal shape has changed (it no longer has `style.typography.fontFamily` property).

I fear that it can introduce subtle bugs. For example, removing block support for font-family should render as invalid blocks the user has set font-family for. I've tried to test this but there're a few moving pieces that make this difficult to replicate in practice: 1) if a block has "more classes" than it should, we still consider it valid ([see](https://github.com/WordPress/gutenberg/blob/96b3508c9e724c231cfee1e694df05f1c9e2e6ee/packages/blocks/src/api/parser/fix-custom-classname.js#L46)), 2) we serialize [all](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/style.js#L212) that it's within the `style` object although the block had removed support for it (as long as the block supports a single style property).

Besides, as soon as the block atts are changed we need to update all the deprecations that use this trick anyway, so it's deferring the changes to the future, when we'll have less context and memory of what happened here.

ockham replied:
Right, I made the mistake of thinking that a deprecation provides attributes for the "resulting" block (post-migration) -- which, of course, now that I think about it, doesn't make sense.

I'm fine reverting these changes 👍

oandregal replied:
Follow-up:

- For the `class` serialization, I think we're bound to keep it working that way (the unrecognized classes are considered "custom classes" added by the user).

- Question about `style` serialization https://github.com/WordPress/gutenberg/issues/35727 I'm not sure we can change this either.

ockham committed:
Un-Express Button block latest deprecation through block.json and save

Essentially reverts 78b2bf518b9f64c2bbe14085943729fcaf3ded0e

ockham committed:
Revert "Express list deprecation through current block.json and save"

This reverts commit 26c57bcff35ed5c77219cb57c16d44791af6e50b.

ockham committed:
List deprecation: Add __experimentalSlashInserter attribute

ockham committed:
Revert "Express PostTitle block deprecation attributes through block.json"

This reverts commit c75c849bb217ec3b37867f5ceb7e2ee04b466ff1.

ockham committed:
Un-Express Navigation block latest deprecation through block.json and save

Essentially reverting f2fd6acf67d5c91f2c6afd1b4389156922063495

ockham committed:
Un-Express Query Title block deprecation attributes through block.json

Essentially reverting db0d9420b29d95fd9bde3537f54240b3b4b3fb39

ockham committed:
Un-Express Site Tagline block deprecation attributes through block.json

Essentially reverting 8c5510d68fe8d23ba7070dabc4c76feadb2bd93d

ockham committed:
Un-Express Site Title block deprecation attributes through block.json

Essentially reverting 6faf83df2bb1a69c7472e505d498cb6304002212

ockham committed:
Un-Express Verse block v2 deprecation through block.json and save

Essentially reverting 35e38a7f32247b50688bb327172607d19e0798e8

oandregal committed:
Make sure slugs are kebab-cased properly

oandregal commented:
Pushed a fix for class serialization at https://github.com/WordPress/gutenberg/pull/31910/commits/21e6ab1847aebe4265ab4e7a2b987feda2a94dc3 See https://github.com/WordPress/gutenberg/pull/35751

oandregal committed:
Update comment

oandregal committed:
Post title: update comment

oandregal committed:
Query title: update comment

oandregal committed:
Site tagline: update comment

oandregal committed:
Site title: update comment

oandregal committed:
Verse: update comment

oandregal committed:
Button: extract font-family deprecation to own constant

oandregal committed:
Button: rename fixtures

oandregal committed:
Button: add font family migration back to other ones

oandregal commented:
Updated the older deprecations for the button block. This is what I'm doing: given the font family was merged in November 2020, I've updated any deprecation older than that date.

I've tested that this button (for example create a file locally at `test/integration/fixtures/blocks/core__button__deprecated-v10-v9.html`):

```html
<!-- wp:button {"borderRadius":25,"style":{"typography":{"fontFamily":"var:preset|font-family|cambria-georgia"}}} -->
<div class="wp-block-button" style="font-family:var(--wp--preset--font-family--cambria-georgia)"><a class="wp-block-button__link" style="border-radius:25px">My button</a></div>
<!-- /wp:button -->
```

is migrated to (execute `npm run fixtures:regenerate test/integration/full-content/full-content.test.js` to get the file `test/integration/fixtures/blocks/core__button__deprecated-v10-v9.serialized.html`):

```html
<!-- wp:button {"fontFamily":"cambria-georgia","style":{"border":{"radius":"25px"}}} -->
<div class="wp-block-button has-cambria-georgia-font-family"><a class="wp-block-button__link" style="border-radius:25px">My button</a></div>
<!-- /wp:button -->
```

I'm not pushing that test to this PR as we'll arguably need many more other tests for each deprecation combination, which can get unwieldy.

Next step: do the same for the navigation block.

oandregal committed:
Navigation: extract deprecation and add missing block supports

oandregal committed:
Navigation: rename test for better consistency

oandregal committed:
Navigation: add font family migration to older ones

oandregal committed:
Make deprecation safe to discard

oandregal committed:
Post Comment Author

oandregal committed:
Post Comment Date

oandregal committed:
Post Date

oandregal commented:
Updated the older deprecations for the navigation block. This is what I'm doing: given the font family was merged in November 2020, I've updated any deprecation older than that date.

I've tested that this navigation block (for example create a file locally at test/integration/fixtures/blocks/core__navigation__deprecated-v4-v3.html):

```html
<!-- wp:navigation {"isResponsive":true,"style":{"typography":{"fontFamily":"var:preset|font-family|cambria-georgia"}}} -->
<!-- wp:navigation-link {"label":"WordPress","url":"https://www.wordpress.org/"} /-->
<!-- /wp:navigation -->
```

is migrated to (execute `npm run fixtures:regenerate test/integration/full-content/full-content.test.js` to get the file `test/integration/fixtures/blocks/core__navigation__deprecated-v4-v3.serialized.html`):

```html
<!-- wp:navigation {"overlayMenu":"mobile","fontFamily":"cambria-georgia"} -->
<!-- wp:navigation-link {"label":"WordPress","url":"https://www.wordpress.org/"} /-->
<!-- /wp:navigation -->
```

I'm not pushing that test to this PR as we'll arguably need many more other tests for each deprecation combination, which can get unwieldy.

oandregal reviewed (approved):
I've tested this manually by adding the blocks in `trunk`, set a font family for them, and then switching to this branch. All were migrated properly.

This is ready! I'll still appreciate a double check by @ockham given this has been a joint effort :)

ockham commented on lib/block-supports/typography.php:
```diff
@@ -105,16 +105,21 @@ function gutenberg_apply_typography_support( $block_type, $block_attributes ) {
+			if ( strpos( $font_family_custom, 'var:preset|font-family' ) !== false ) {
+				$index_to_splice  = strrpos( $font_family_custom, '|' ) + 1;
+				$font_family_slug = substr( $font_family_custom, $index_to_splice );
+				$classes[]        = sprintf( 'has-%s-font-family', $font_family_slug );
```
Do we need `gutenberg_experimental_to_kebab_case` here?

```suggestion
				$classes[]        = sprintf( 'has-%s-font-family', gutenberg_experimental_to_kebab_case( $font_family_slug ) );
```

oandregal replied:
This was actually a bug in the old code path. I've added tests for the three code paths and fixed this at https://github.com/WordPress/gutenberg/pull/31910/commits/19dbeee39b119ec8e814e7b5bd9200f4798e9a3d

ockham commented:
> Updated the older deprecations for the button block. This is what I'm doing: given the font family was merged in November 2020, I've updated any deprecation older than that date.

Not sure I'm totally getting this one 😅 

Don't we need to update all button block deprecations _younger_ than a given cutoff date (rather than older), in order to catch the ones that were introduced during the time that the font family related attributes have been around?

Also, I'm not sure about the cutoff date? I see `migrateFontFamily` in deprecations that seem to have been added in [September 2019](https://github.com/WordPress/gutenberg/pull/31910/files#diff-e133a7448e28a248ec35cb9a2943bb23e00755612533ac29e015532a6d618f9eR794) or [one that seems to have been added in April 2020](https://github.com/WordPress/gutenberg/pull/31910/files#diff-e133a7448e28a248ec35cb9a2943bb23e00755612533ac29e015532a6d618f9eR677), but not [another](https://github.com/WordPress/gutenberg/pull/31910/files#diff-e133a7448e28a248ec35cb9a2943bb23e00755612533ac29e015532a6d618f9eR873) 🤔  (judging by the commit dates of the surrounding lines according to Git Lens).

oandregal committed:
Add tests and fix legacy code path

oandregal committed:
Button: revert unnecessary font family migration

oandregal committed:
Button: revert unnecessary font family migration

oandregal committed:
Button: revert unnecessary font family migration

oandregal commented:
> Don't we need to update all button block deprecations younger than a given cutoff date (rather than older), in order to catch the ones that were introduced during the time that the font family related attributes have been around?

That's correct! It was also what I meant but my English skills are failing me, apparently :angel: 

> Also, I'm not sure about the cutoff date? I see migrateFontFamily in deprecations that seem to have been added in 
> September 2019 or one that seems to have been added in April 2020, but not another thinking (judging by the commit dates of the surrounding lines according to Git Lens).

Great catch. I was misguided because I checked the lines of those migrations that were actually modified later (to add newer migrations).

ockham committed:
Post Title block deprecation: Add spacing supports (sync from block.json)

ockham committed:
Query Title block deprecation: Add spacing supports (sync from block.json)

ockham reviewed (commented):
I went through all newly modified files again (and added two attributes to two deprecations that have been added very recently, i.e. in the past few hours), and verified that the composed deprecations work, as described [here](https://github.com/WordPress/gutenberg/pull/31910#issuecomment-946883202) and [here](https://github.com/WordPress/gutenberg/pull/31910#issuecomment-947463456).

I didn't do another round of manually testing all block deprecations since I think those are covered fine by the fixtures.

Anyway, LGTML! Let's land this 👍 ✨ 🚢 💥
